From 63bed294f4075e3f1aae5f1c8136211564236fc9 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Thu, 20 Mar 2025 15:35:32 +0100
Subject: [PATCH 1652/1695] clk: stm32mp21: don't put all flexgen clocks as
 critical

No longer necessary to mark all the flexgen clocks as critical for
the coretx-M33 because it now uses the SCMI protocol.

Only keep as critical the root clock (flexgen 0 to 5) and flexgen 63
for DVFS support (CA35 bypass).

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git a3b087373cbe192b202542ee9cd9fecbe5dde413 clk: stm32mp21: don't put all flexgen clocks as critical]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: I6996ac4177ee25b72e206af7d1195ce4462dfbb0
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/445675
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/448890
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 core/drivers/clk/clk-stm32mp21.c | 39 ++++++++++++--------------------
 1 file changed, 14 insertions(+), 25 deletions(-)

diff --git a/core/drivers/clk/clk-stm32mp21.c b/core/drivers/clk/clk-stm32mp21.c
index 2fcdfa4f4..2182c11a5 100644
--- a/core/drivers/clk/clk-stm32mp21.c
+++ b/core/drivers/clk/clk-stm32mp21.c
@@ -4139,34 +4139,25 @@ static struct clk *stm32mp21_clk_provided[STM32MP21_ALL_CLK_NB] = {
 	[CK_MSI_KER]		= &ck_msi_ker,
 };
 
-static void clk_stm32_set_flexgen_as_critical(void)
-{
-	uint32_t i = 0;
-
-	for (i = 0; i < XBAR_CHANNEL_NB; i++) {
-		unsigned long clock_id = CK_ICN_HS_MCU + i;
-		struct clk *clk = NULL;
-
-		if (!stm32_rcc_has_access_by_id(i) || i == 6)
-			continue;
-
-		clk = stm32mp_rcc_clock_id_to_clk(clock_id);
-		assert(clk);
-
-		clk_enable(clk);
-	}
-}
-
 static bool clk_stm32_clock_is_critical(struct clk *clk __maybe_unused)
 {
-#ifndef CFG_STM32_CM33TDCID
-
 	struct clk *clk_criticals[] = {
+#ifdef CFG_STM32_CM33TDCID
+		&ck_flexgen_63,
+#else /* CFG_STM32_CM33TDCID */
 		&ck_hsi,
 		&ck_hse,
 		&ck_msi,
 		&ck_lsi,
 		&ck_lse,
+		&ck_icn_hs_mcu,
+		&ck_icn_ls_mcu,
+		&ck_icn_sdmmc,
+		&ck_icn_ddr,
+		&ck_icn_display,
+		&ck_icn_hsl,
+		&ck_icn_nic,
+		&ck_flexgen_63,
 		&ck_cpu1,
 		&ck_icn_p_syscpu1,
 		&ck_icn_s_ddr,
@@ -4190,7 +4181,8 @@ static bool clk_stm32_clock_is_critical(struct clk *clk __maybe_unused)
 		&ck_icn_p_gpioh,
 		&ck_icn_p_gpioi,
 		&ck_icn_p_gpioz,
-		&ck_icn_p_ipcc1
+		&ck_icn_p_ipcc1,
+#endif /* CFG_STM32_CM33TDCID */
 	};
 	size_t i = 0;
 
@@ -4200,7 +4192,7 @@ static bool clk_stm32_clock_is_critical(struct clk *clk __maybe_unused)
 		if (clk == clk_critical)
 			return true;
 	}
-#endif /* CFG_STM32_CM33TDCID */
+
 	return false;
 }
 
@@ -4627,9 +4619,6 @@ end:
 		}
 	}
 
-	/* TEMPORARY: Enables all root clocks (could be used by M33) */
-	clk_stm32_set_flexgen_as_critical();
-
 	return TEE_SUCCESS;
 }
 
-- 
2.47.2

